gotrue/types.py

```diff
@@ -0,0 +1,344 @@
+from dataclasses import dataclass
+from enum import Enum
+from inspect import signature
+from json import dumps
+from time import time
+from typing import Any, Callable, Dict, List, Optional, Type, TypeVar
+
+T = TypeVar("T")
+
+
+def parse_none(v: Optional[T], f: Callable[[Any], T]) -> Optional[T]:
+    if v is None:
+        return None
+    return f(v)
+
+
+def parse_dict(cls: Type[T], **json: dict) -> T:
+    cls_fields = {field for field in signature(cls).parameters}
+    native_args, new_args = {}, {}
+    for name, val in json.items():
+        if name in cls_fields:
+            native_args[name] = val
+        else:
+            new_args[name] = val
+    ret = cls(**native_args)
+    for new_name, new_val in new_args.items():
+        setattr(ret, new_name, new_val)
+    return ret
+
+
+@dataclass
+class ApiError(BaseException):
```

Comment:
```suggestion
class APIError(BaseException):
```
[PEP-8 says that when you use acronyms in PascalCase names, the acronym has to be in all-caps.](https://www.python.org/dev/peps/pep-0008/#descriptive-naming-styles) Please fix this in other spots as well.

Reply:
Ok 👍🏼